### PR TITLE
TSPS-305 Add GCS service, sign PUT urls for data uploads

### PR DESCRIPTION
**Pull request by mmorgantaylor**
### Description 

Note: #118 contains updates that affect this PR; if that one is merged first, will need to update this PR.

The result of preparePipelineRun is now:
```{
  "jobId": "ffaffa12-5717-4562-b3fc-2c963f66afa6",
  "fileInputUploadUrls": {
    "multiSampleVcf": {
      "signedUrl": "https://storage.googleapis.com/fc-secure-6970c3a9-dc92-436d-af3d-917bcb4cf05a/user-input-files/ffaffa12-5717-4562-b3fc-2c963f66afa6/TEST.vcf.gz?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=tsps-dev%40broad-dsde-dev.iam.gserviceaccount.com%2F20240823%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20240823T170006Z&X-Goog-Expires=900&X-Goog-SignedHeaders=content-type%3Bhost&X-Goog-Signature=[redacted]",
      "curlCommand": "curl -X PUT -H 'Content-Type: application/octet-stream' --upload-file TEST.vcf.gz 'https://storage.googleapis.com/fc-secure-6970c3a9-dc92-436d-af3d-917bcb4cf05a/user-input-files/ffaffa12-5717-4562-b3fc-2c963f66afa6/TEST.vcf.gz?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=tsps-dev%40broad-dsde-dev.iam.gserviceaccount.com%2F20240823%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20240823T170006Z&X-Goog-Expires=900&X-Goog-SignedHeaders=content-type%3Bhost&X-Goog-Signature=[redacted]'"
    }
  }
}
```

### Jira Ticket
https://broadworkbench.atlassian.net/browse/TSPS-305

### CODE REVIEW

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import com.google.cloud.storage.StorageOptions;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GcsClient {
+
+  public GcsClient() {}
+
+  public Storage getStorageService(String projectId) {
+    // this will use application default credentials
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java:13]: we should probably explicitly use the SA creds, right?

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucketName, objectName)).build();
+
+    // Generate Signed URL
+    Map<String, String> extensionHeaders = new HashMap<>();
+    extensionHeaders.put("Content-Type", "application/octet-stream");
+
+    URL url =
+        storage.signUrl(
+            blobInfo,
+            15,
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java:56]: need to update these values - put in a config

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package bio.terra.pipelines.dependencies.gcs;
+
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GcsClient {
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java:8]: wasn't totally sure what should go into the client vs service on this one

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java [hunk 4]
```diff
@@ -109,6 +121,55 @@ public void preparePipelineRun(
...
+    List<String> fileInputNames =
+        pipeline.getPipelineInputDefinitions().stream()
+            .filter(PipelineInputDefinition::getUserProvided)
+            .filter(p -> p.getType().equals(PipelineVariableTypesEnum.FILE))
+            .map(PipelineInputDefinition::getName)
+            .toList();
+
+    // TODO don't hardcode this
+    String projectId = "terra-dev-244bacba";
+    String bucketName = pipeline.getWorkspaceStorageContainerUrl().replace("gs://", "");
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:151]: should we remove the `gs://` prefix here, or should we store the bucket name without the `gs://`?

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java [hunk 5]
```diff
@@ -109,6 +121,55 @@ public void preparePipelineRun(
...
+    // get the list of files that the user needs to upload
+    List<String> fileInputNames =
+        pipeline.getPipelineInputDefinitions().stream()
+            .filter(PipelineInputDefinition::getUserProvided)
+            .filter(p -> p.getType().equals(PipelineVariableTypesEnum.FILE))
+            .map(PipelineInputDefinition::getName)
+            .toList();
+
+    // TODO don't hardcode this
+    String projectId = "terra-dev-244bacba";
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:150]: store this value in db along with other workspace params? (this is specific to the workspace)

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java [hunk 6]
```diff
@@ -97,6 +105,10 @@ public void preparePipelineRun(
               .formatted(jobId));
     }
 
+    // return a map of SAS urls and azcopy commands for the user to upload their input files
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:108]: update comment

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+  private final GcsClient gcsClient;
+
+  private static final Logger logger = LoggerFactory.getLogger(GcsService.class);
+
+  public GcsService(GcsClient gcsClient) {
+    this.gcsClient = gcsClient;
+  }
+
+  /**
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java:28]: update this docstring

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java (hunk 3 above)
Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java:8]: i think having this client is a good separation.  this should be simple and have no business logic in it

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java (hunk 1 above)
Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsClient.java:13]: application default credentials are set as the SA creds already, we do that [here](https://github.com/DataBiosphere/terra-scientific-pipelines-service/blob/main/service/build.gradle#L118)

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    // Define Resource
+    BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucketName, objectName)).build();
+
+    // Generate Signed URL
+    Map<String, String> extensionHeaders = new HashMap<>();
+    extensionHeaders.put("Content-Type", "application/octet-stream");
+
+    URL url =
+        storage.signUrl(
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java:54]: we probably wnat to wrap this in a similar retry mechanism as we do our other services.  you might have to dig a bit to see hwat all exceptions exist or error modes

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java (hunk 5 above)
Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:150]: ughhh the amountof things we're storing for a workspace is annoying.  Is this available through a terra api.  Maybe this is one of those we query for when we hit our Admin API

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java (hunk 4 above)
Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:151]: is we're calling it `Url` then it shoudl have gs://.  we could call it just `Name` instead and then use it appropriately in other places

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java (hunk 5 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:150]: yep, it is available via terra api (rawls: https://rawls.dsde-prod.broadinstitute.org/#/workspaces/listWorkspaceDetails you get workspace.googleProject and workspace.bucketName). i can make a ticket to retrieve both those things when you call admin endpoint

File: service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java (hunk 8 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/dependencies/gcs/GcsService.java:54]: thanks, yes, i did mean to come back and add retries

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java (hunk 5 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:150]: https://broadworkbench.atlassian.net/browse/TSPS-307

File: service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java (hunk 4 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:151]: i like that change - added that to AC for TSPS-307

Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java:151]: if https://github.com/DataBiosphere/terra-scientific-pipelines-service/pull/118 is merged first, will need to update this

**Review by jsotobroad — APPROVED** [review]
looks good

**Pull request merged**